### DISCUSSION TITLE
Is the REST API for uploading images documented anywhere?

### DISCUSSION THREAD

**Discussion by dazzag24** [requarks/wiki#6049] [Guides]
Related to #2746 I would like to know if there is any documentation for the REST API for uploading image assets to Wiki.js?

The GraphQL API does not provide the ability to upload images, which means you cannot create pages that include images using the  GraphQL API.  In #2746 I was told that you could use the REST API to upload images, but there is no documentation for this that I have been able to find.

Thanks

Darren

**Comment by theoratkin** [comment]
I struggled with it as well, until I came upon https://github.com/Requarks/wiki/issues/2413#issuecomment-689876881. But it doesn't mention that you have to specify `folderId` as well, and lacks any examples. I guess it's easy enough to figure out for most developers, but I'm not as experienced with HTTP requests, and I had to do lots of trial and error to get it working.

Until it is properly documented, here's a description for everyone who might need it:

Send a multipart/form-data POST request to `/u` with two parts (order doesn't matter), both named `mediaUpload`: the first part is a JSON string describing the `folderId` value (`0` is the root folder), the second part is image data. Here's an example in cURL:

```sh
curl 'http://your.wiki.com/u' \
	-H 'Authorization: Bearer API_KEY' \
	-F 'mediaUpload={"folderId":0}' \
	-F 'mediaUpload=@image.jpg'
```

And here's Python `requests` example:

```python
import requests

upload_data = (
    ('mediaUpload', (None, '{"folderId":0}')),
    ('mediaUpload', ('image.jpg', open('image.jpg', 'rb'), 'image/jpeg'))
)

response = requests.post('http://your.wiki.com/u',
        headers={'Authorization' : 'Bearer API_KEY'},
        files=upload_data)

print(response.content)
```

And here's what it would translate into:

```
-----------------------------263214647317036536961330815606
Content-Disposition: form-data; name="mediaUpload"

{"folderId":0}
-----------------------------263214647317036536961330815606
Content-Disposition: form-data; name="mediaUpload"; filename="image.jpg"
Content-Type: image/jpeg

Image data....
```

**Comment by yourjelly** [comment]
I'm trying to upload with curl following your example here, but I'm getting a 'Missing upload folder metadata." error, that has stopped me.
`{"succeeded":false,"message":"Missing upload folder metadata."}`

Does anyone have an idea for what I could be missing?

**Reply by theoratkin** [reply]
Hey, sorry, it's been a few years since I've used Wiki.js. The API might have changed.

Is the image uploading API still not documented anywhere? I'm really surprised if that's the case.

Sadly I don't think I can offer a solution, but you could try inspecting the upload request via a browser, it should give you a clue of what exactly is missing.

**Reply by yourjelly** [reply]
The solution for me is I needed to escape the quotations on folderId, dont know if it was because of powershell or what.
```
curl.exe 'http://your.wiki.com/u' \
	-H 'Authorization: Bearer API_KEY' \
	-F 'mediaUpload={/"folderId/":0}' \
	-F 'mediaUpload=@image.jpg'
```
